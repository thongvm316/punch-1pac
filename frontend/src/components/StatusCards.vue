<template>
  <div class="status-cards mt-5">
    <datepicker
      :language="currentUser.language"
      :format="function (date) { return $moment(date).format('LLL') }"
      :minimumView="'month'"
      :maximumView="'month'"
      :input-class="'datepicker-input form-input'"
      :calendar-class="'datepicker-calendar'"
      :wrapper-class="'datepicker'"
      @input="onInputDatepicker"
      v-model="month"/>
    <div class="columns">
      <div class="column col-4" v-for="(status, key) in i18nStatuses" :key="key">
        <div class="box mt-5">
          <p>{{ $t(`meta.attendance_statuses.${status}`) }}</p>
          <h3>{{ $tc('statusCards.dayNum', statuses[status], { num: statuses[status], companyTotalDays: meta.company_total_working_days_in_month }) }}</h3>
        </div>
      </div>
      <div class="column col-4" v-if="statuses['working_hours']">
        <div class="box mt-5">
          <p>{{ $t('meta.attendance_statuses.working_hours') }}</p>
          <h3>{{ $t('statusCards.workingHours', { hours: statuses['working_hours'].hours, mins: statuses['working_hours'].mins, companyTotalHours: meta.company_total_working_hours_on_month }) }}</h3>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import { mapState, mapActions } from 'vuex'
const Datepicker = () => import('vuejs-datepicker')

export default {
  name: 'status-cards',

  data() {
    return {
      month: this.$moment()
        .locale('en')
        .format('LL')
    }
  },

  components: {
    Datepicker
  },

  computed: {
    ...mapState('statusCards', ['statuses', 'meta']),

    i18nStatuses() {
      return Object.keys(this.statuses).filter(status => !['working_hours', 'minutes_attend_late', 'minutes_leave_early'].includes(status))
    }
  },

  methods: {
    ...mapActions('statusCards', ['getStatuses']),

    onInputDatepicker() {
      this.month = this.$moment(this.month).format('YYYY-MM-DD')
    }
  },

  created() {
    this.getStatuses(this.month)
  },

  watch: {
    month: function() {
      this.getStatuses(this.month)
    }
  }
}
</script>
