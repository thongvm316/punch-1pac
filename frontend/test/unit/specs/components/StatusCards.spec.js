import { shallowMount } from '@vue/test-utils'

import wrapperOps from '../../supports/wrapper'
import localVue from '../../supports/local-vue'

import Datepicker from 'vuejs-datepicker'
import StatusCards from '@/components/StatusCards'

const onInputDatepicker = jest.spyOn(StatusCards.methods, 'onInputDatepicker')
const month = localVue.prototype.$moment().format('LL')
Object.assign(wrapperOps, {
  data: function() {
    return {
      month
    }
  },
  methods: { onInputDatepicker }
})

describe('StatusCards.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(StatusCards, wrapperOps)
  })

  afterEach(() => { wrapper.vm.$destroy() })

  it('should render correctly', () => {
    expect(wrapper.exists()).toBeTruthy()
    expect(wrapper.isVueInstance()).toBeTruthy()
    expect(wrapper.find(Datepicker).exists()).toBeTruthy()
  })

  it('should return onInputDatePicker', () => {
    wrapper.vm.onInputDatepicker()
    const localMonth = wrapper.vm.month
    expect(localMonth).toEqual(localVue.prototype.$moment(localMonth).format('YYYY-MM-DD'))
  })
})
