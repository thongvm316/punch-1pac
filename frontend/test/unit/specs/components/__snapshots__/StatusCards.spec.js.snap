// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusCards.vue should render correctly 1`] = `
<div class="status-cards mt-5">
  <datepicker-stub language="en" format="MMMM yyyy" minimum-view="month" maximum-view="month" input-class="datepicker-input form-input" calendar-class="datepicker-calendar" wrapper-class="datepicker" value="Feb 05, 2019"></datepicker-stub>
  <div class="columns">
    <div class="column col-4">
      <div class="box mt-5">
        <p>Attend Late</p>
        <h3>8 / 20 days</h3>
      </div>
    </div>
    <div class="column col-4">
      <div class="box mt-5">
        <p>Attend OK</p>
        <h3>14 / 20 days</h3>
      </div>
    </div>
    <div class="column col-4">
      <div class="box mt-5">
        <p>Day Off</p>
        <h3>0 / 20 days</h3>
      </div>
    </div>
    <div class="column col-4">
      <div class="box mt-5">
        <p>Leave Early</p>
        <h3>12 / 20 days</h3>
      </div>
    </div>
    <div class="column col-4">
      <div class="box mt-5">
        <p>Leave OK</p>
        <h3>10 / 20 days</h3>
      </div>
    </div>
    <div class="column col-4">
      <div class="box mt-5">
        <p>Working hours</p>
        <h3>0h0m / 160h</h3>
      </div>
    </div>
  </div>
</div>
`;
